dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -25,260 +26,293 @@
-
-		private final String handlerId;
-		private final String sourceId;
-		private long seqNum;
-
-		// Variables to keep track of lost events
-		private long lostEventsForTrace;
-		private long lostEventsForWarning;
-		private long totalLostEvents;
-
-		private long lastReportTimeForTrace;
-		private long lastReportTimeForWarning;
-		private final long intervalForTrace = 60 * 1000;
-		private final long intervalForWarning = 60 * 5 * 1000;
-
-		public HandlerStats(String handlerId, String sourceId) {
-			this.handlerId = handlerId;
-			this.sourceId = sourceId;
-			seqNum = 1;
-			lostEventsForWarning = lostEventsForTrace = totalLostEvents = 0;
-			lastReportTimeForWarning = 0;
-			lastReportTimeForTrace = System.currentTimeMillis();
-		}
-
-		public long getNextSeqNum() {
-			return seqNum;
-		}
-
-		public void setNextSeqNum(long nextSeqNum) {
-			this.seqNum = nextSeqNum;
-		}
-
-		public void traceEventLoss(long retSeqNum) {
-			if (retSeqNum > seqNum) {
-				long eventsLost = retSeqNum - seqNum;
-				lostEventsForWarning += eventsLost;
-				lostEventsForTrace += eventsLost;
-				totalLostEvents += eventsLost;
-			}
-			long currentTime = System.currentTimeMillis();
-			long timeElapsed = currentTime - lastReportTimeForWarning;
-			if (lostEventsForWarning > 0 && timeElapsed >= intervalForWarning) {
-				if (lastReportTimeForWarning == 0) {
-					// This was the first instance where we started to see loss
-					// of events.
-					Tr.warning(tc, "HANDLER_STARTED_TO_LOSE_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId);
-				} else {
-					long timeElapsedInMins = TimeUnit.MILLISECONDS.toMinutes(timeElapsed);
-					Tr.warning(tc, "HANDLER_LOST_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId,
-							timeElapsedInMins, totalLostEvents);
-				}
-				lastReportTimeForWarning = currentTime;
-				lostEventsForWarning = 0;
-			}
-			timeElapsed = currentTime - lastReportTimeForTrace;
-			if (timeElapsed >= intervalForTrace) {
-				if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
-					long timeElapsedInSecs = TimeUnit.MILLISECONDS.toSeconds(timeElapsed);
-					Tr.event(tc,
-							"Handler [{0}] has lost {1} events from source [{2}] in the last {3} second(s), and has lost {4} events from the source since the handler started.",
-							handlerId, lostEventsForTrace, sourceId, timeElapsedInSecs, totalLostEvents);
-				}
-				lastReportTimeForTrace = currentTime;
-				lostEventsForTrace = 0;
-			}
-		}
-
-	}
+    private static final ReentrantReadWriteLock RERWLOCK = new ReentrantReadWriteLock(true);
+    private Set<SynchronousHandler> synchronousHandlerSet = new HashSet<SynchronousHandler>();
+    private int capacity;
+
+    private final String sourceId;
+    /* Map to keep track of the next event for a handler */
+    private final ConcurrentHashMap<String, HandlerStats> handlerEventMap = new ConcurrentHashMap<String, HandlerStats>();
+
+    private volatile static boolean EMQRemovedFlag = false;
```

Comment:
remove private from EMQRemovedFlag so you can see it from test classes in same package